shardingsphere-infra/shardingsphere-infra-executor/src/main/java/org/apache/shardingsphere/infra/executor/sql/federate/schema/table/generator/FederateExecutionSQLGenerator.java

```diff
@@ -35,14 +38,17 @@
     
     private final int[] projects;
     
+    private final List<RelDataTypeField> fields;
+    
     /**
      * Generate sql.
      *
      * @param table table
      * @return sql
      */
     public String generate(final String table) {
-        // TODO
-        return String.format("SELECT * FROM %s", table);
+        // TODO generate sql with filters
+        String project = null == projects || 0 == projects.length ? "*" : Arrays.stream(projects).mapToObj(each -> fields.get(each).getName()).collect(Collectors.joining(", "));
```

Comment:
Which condition will make`project = null == projects || 0 == projects.length`  true?

Reply:
> projects

@tristaZero It can be seen from the source code that when the original projects and the optimized projects are the same, projectInts will return null. Maybe the `0 == projects.length` condition is redundant.

```java
  private static TableScanNode createProjectableFilterable(Compiler compiler,
      TableScan rel, ImmutableList<RexNode> filters, ImmutableIntList projects,
      ProjectableFilterableTable pfTable) {
    final DataContext root = compiler.getDataContext();
    final ImmutableIntList originalProjects = projects;
    for (;;) {
      final List<RexNode> mutableFilters = Lists.newArrayList(filters);
      final int[] projectInts;
      if (projects == null
          || projects.equals(TableScan.identity(rel.getTable()))) {
        projectInts = null;
      } else {
        projectInts = projects.toIntArray();
      }
      final Enumerable<Object[]> enumerable1 =
          pfTable.scan(root, mutableFilters, projectInts);
...
}
```